src/python/pants/core/goals/clean.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+import tempfile
+from pathlib import Path
+from typing import Union
+
+import humanize
+
+from pants.base.build_environment import get_pants_cachedir
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.options_parsing import _Options
+from pants.engine.rules import collect_rules, goal_rule
+from pants.option.global_options import LocalStoreOptions
+from pants.util.dirutil import rm_rf
+
+# N.B. We use a list so that named-caches can go last, as it is quite large and for dry-run
+# it's a nicer UX if it's last.
+_NAMED_CACHE_FLAGS_INFO = [
+    (
+        "process-execution-temp-dirs",
+        "Whether to keep the process execution temporary directories. These are normally cleaned up"
+        " automatically, however they are leaked if --no-process-execution-local-cleanup.",
+    ),
+    (
+        "lmdb-store",
+        "Whether to keep the local file store, which stores the results of subprocesses run by Pants.",
+    ),
+    (
+        "run-tracker",
+        "Whether to clean directories persisted after using `./pants run`.",
+    ),
+    (
+        "repl-temp-dirs",
+        "Whether to clean directories persisted after using `./pants repl`.",
+    ),
+    ("dist-dir", "Whether to clean the distributable files directory. "),
+    ("bootstrap-dir", "Whether to clean directories used by pants for bootstrapping itself."),
+    ("named-caches", "Whether to clean the named global caches."),
+]
+
+
+def _dir_info(*paths: Path):
+    total_size = num_files = 0
+    for path in paths:
+        for dirpath, _, filenames in os.walk(path):
+            total_size += sum(os.path.getsize(os.path.join(dirpath, f)) for f in filenames)
+            num_files += len(filenames)
+    return total_size, num_files
+
+
+class CleanSubsystem(GoalSubsystem):
+    name = "clean"
+    help = "Clean pants caches and temp dirs (based on current settings)"
+
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--dry-run",
+            advanced=False,
+            type=bool,
+            default=True,
+            help=("Whether to just report what would be cleaned. @TODO: more"),
+        )
+        register(
+            "--all",
+            advanced=False,
+            type=bool,
+            default=False,  # @TODO: True?
```

Comment:
I don't think so... `clean` is a very large hammer, and I think that one of the goals for this ... goal... should be to have very useful help (which we can continue to expand after landing!) that explains what each cache is for, and _why_ you might want to clean it. Most of these will effectively never need to be cleaned, but when you do clean them, you set yourself up for tens of minutes of building and refetching.

When people are experiencing caching bugs, you could almost imagine the help strings adding up to a sort of "wizard" that walks you through what to try. See https://www.pantsbuild.org/docs/troubleshooting#cache-or-pantsd-invalidation-issues for some more on this (and again, we can expand this after landing).